### ISSUE TITLE
Models refactor

### ISSUE THREAD

**Issue by lgvaz** [lgvaz/mantisshrimp#137]
## Overview

As @ai-fast-track suggested, we could use a more functional approach for the methods currently present in `MantisModule` and more specifically in `MantisRCNN`.

The main source of confusion is that our models are currently doing two things: Defining the neural network and implementing the algorithm.

To be more specific, these are the methods our models currently require:
* forward: how to feed-forward the NN
* loss_func: given the preds and targets calculate the loss
* dataloader: given the dataset get the dataloader
* build_training_sample: convert from a prepared_record to something the model can understand. This works as an adapter between our data structures and the data structures expected by the NN.

The main point of discussion was the `.dataloader` method, claiming that it's not logical to have a method on the model that returns a `DataLoader`, and it's more logical to just have a function like "rcnn_dataloader" that return a `DataLoader` with the required modifications (like a custom collate_fn).

## Objective of the interface
Provide functions that ease out the integration of models/algorithms with other training frameworks. For example, any training framework will require the specification of a loss function, our library should be able to provide that.

---
**Note:**
Currently we're only working with torchvision, and this is biasing us of having everything into the model because this is the way they done it.

It's unnatural to separate those functions because the model is already tightly coupled in their implementation, while training, their model actually requires you to pass the targets to the `forward` method so then return you a list of losses instead of the predictions! (and this is very problematic, for instance, in validation we need to do one forward pass to get the losses and another to get the predictions)

**Suggestion:**
Similar to what we did when making an agnostic training api, I suggest that we first try to integrate another model/algorithm to the current structure before trying to come up with another API.

This would give us another point of view and make things clearer on how a more generic API should look like, forcing us into making better decisions.

Yolo could be a good addition to the library, compared to MaskRCNN, Yolo is an algorithm that trades a bit of precision for a lot of speed. We can also try to integrate Ross efficientdet (although at this moment I don't really understand how efficientdet works)

**Comment by lgvaz** [comment]
This is going to be one of the most important interfaces of mantisshrimp so it has to be very clear and intuitive!

I encourage everyone to give their suggestions so we can settle on an interface that is modular and straight forward! 🔥

**Comment by lgvaz** [comment]
@ai-fast-track , @oke-aditya , @philtrade

**Comment by lgvaz** [comment]
Currently, we have a single function for getting a dataloader, called on `model.dataloader`. It's the case for `RCNN` that the training and validation dataloaders are identical, **but** for efficient-det they are a slightly different...

Do you think it's okay to have 3 functions for dataloaders?
* train_dataloader
* valid_dataloader
* inference_dataloader

In the case train and valid are equal (for RCNN), we just need to call `train_dataloader` inside `valid_dataloader` for example.

@ai-fast-track and @oke-aditya , what do you think?

**Comment by philtrade** [comment]
IMHO it's too wide an API to have three dataloader functions.

One way is to allow users to inherit and tweak a standard one into slightly different one for whatever purpose, be it for valid or inference or different sampling strategy.

i.e. allow parameterization on the difference, either by composition (user supplies a sampler, for example), or by inheritance.

**Comment by oke-aditya** [comment]
We can have 2 data loaders. `train_dataloader` like `model.train_loader` and `validation_loader` like `model.valid_loader`. Users can use the same valid loader as test loader too. I see no big difference. But as Phile Suggested if the user want a custom dataloader he can inherit from `dataloader` and implement one himself.

**Comment by lgvaz** [comment]
First of all, I'm also not a big fan of having three dataloaders (or two for that matter), this is why the original design included only one. That being said, the use cases are forcing us to distinguish, let me tell exactly why:

What really changes, is how you build a training sample, and more specifically that happens inside the `collate_fn` of a dataloader. Let's take efficientdet as an example:

- Train sample: A batch has to contain two things, images and targets. Images is a tensor of shape (bs, c, w, h) and targets is a dictionary with the keys `cls` and `labels`
- Valid sample: Very similar to the train, but target also needs the keys `img_scale` and `img_size`
- Inference sample: A batch only contain the images.

To the suggestions of having only two dataloaders (`train` and `valid`) I hope you can see why that it's not possible, `valid` and `inference` are very different, in the sense that `valid` needs targets and `inference` don't.

Basically, what really change is the `collate_fn`, we could eliminate the need for any custom dataloaders by simply supplying the `collate_fn` to the user, but that would just sidestep the issue, we would still have three `collate_fn`

@philtrade It's also worth saying that we're never inheriting from a `DataLoader`, the only thing that we do is pre-saturate them with custom arguments (like the specific `collate_fn` for a model)
